File: basic/old/sample-objc/SendBird-iOS/UserProfile/UserProfileViewController.m
//
//  UserProfileViewController.m
//  SendBird-iOS
//
//  Created by Jed Gyeong on 3/27/17.
//  Copyright © 2017 SendBird. All rights reserved.
//

#import <SendBirdSDK/SendBirdSDK.h>
#import <AFNetworking/UIImageView+AFNetworking.h>
#import <AFNetworking/AFImageDownloader.h>
#import <MobileCoreServices/UTCoreTypes.h>
#import <MobileCoreServices/UTType.h>
#import <Photos/Photos.h>

#import "UserProfileViewController.h"
#import "Constants.h"
#import "NSBundle+SendBird.h"


@interface UserProfileViewController ()

@property (weak, nonatomic) IBOutlet UINavigationItem *navItem;
@property (weak, nonatomic) IBOutlet UIImageView *profileImageView;
@property (weak, nonatomic) IBOutlet UITextField *nicknameTextField;
@property (weak, nonatomic) IBOutlet NSLayoutConstraint *bottomMargin;
@property (weak, nonatomic) IBOutlet UISwitch *pushNotificationSwitch;
@property (weak, nonatomic) IBOutlet UIActivityIndicatorView *updatingIndicator;

@property (strong) NSData *profileImageData;

@end

@implementation UserProfileViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view from its nib.
    UIBarButtonItem *negativeRightSpacer = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace target:nil action:nil];
    negativeRightSpacer.width = -2;
    UIBarButtonItem *rightDisconnectItem = [[UIBarButtonItem alloc] initWithTitle:@"Save" style:UIBarButtonItemStylePlain target:self action:@selector(save)];
    [rightDisconnectItem setTitleTextAttributes:@{NSFontAttributeName: [Constants navigationBarButtonItemFont]} forState:UIControlStateNormal];
    
    UIBarButtonItem *negativeLeftSpacer = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace target:nil action:nil];
    negativeLeftSpacer.width = -2;
    UIBarButtonItem *leftProfileItem = [[UIBarButtonItem alloc] initWithTitle:@"Close" style:UIBarButtonItemStylePlain target:self action:@selector(close)];
    [leftProfileItem setTitleTextAttributes:@{NSFontAttributeName: [Constants navigationBarButtonItemFont]} forState:UIControlStateNormal];
    
    self.navItem.rightBarButtonItems = @[negativeRightSpacer, rightDisconnectItem];
    self.navItem.leftBarButtonItems = @[negativeLeftSpacer, leftProfileItem];
    
    [self.profileImageView setImageWithURL:[NSURL URLWithString:[SBDMain getCurrentUser].profileUrl]];
    
    self.profileImageData = nil;
    
    self.updatingIndicator.hidden = YES;
    
    UITapGestureRecognizer *profileImageTapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(clickProfileImage)];
    self.profileImageView.userInteractionEnabled = YES;
    [self.profileImageView addGestureRecognizer:profileImageTapRecognizer];
    
    [self.nicknameTextField setText:[SBDMain getCurrentUser].nickname];
    
    BOOL isRegisteredForRemoteNotifications = [[UIApplication sharedApplication] isRegisteredForRemoteNotifications];
    [self.pushNotificationSwitch setOn:isRegisteredForRemoteNotifications];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)close {
    [self dismissViewControllerAnimated:NO completion:nil];
}

- (void)save {
    if ([self.nicknameTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]].length == 0) {
        return;
    }
    
    self.updatingIndicator.hidden = NO;
    [self.updatingIndicator startAnimating];
    
    BOOL isRegisteredForRemoteNotifications = [self.pushNotificationSwitch isOn];
    if (isRegisteredForRemoteNotifications) {
        [[UIApplication sharedApplication] registerForRemoteNotifications];
        [SBDMain updateCurrentUserInfoWithNickname:[self.nicknameTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] profileImage:self.profileImageData completionHandler:^(SBDError * _Nullable error) {
            BOOL hasUpdatedUserInfo;
            if (error != nil) {
                hasUpdatedUserInfo = NO;
            }
            else {
                AFImageDownloader *imageDownloader = [AFImageDownloader defaultInstance];
                NSURLCache *urlCache = imageDownloader.sessionManager.session.configuration.URLCache;
                [urlCache removeAllCachedResponses];
                [imageDownloader.imageCache removeImageWithIdentifier:[SBDMain getCurrentUser].profileUrl];
                
                hasUpdatedUserInfo = YES;
            }

#if !(TARGET_OS_SIMULATOR)
            [SBDMain registerDevicePushToken:[SBDMain getPendingPushToken] unique:YES completionHandler:^(SBDPushTokenRegistrationStatus status, SBDError * _Nullable error) {
                BOOL hasUpdatedPushNoti;
                
                if (error != nil) {
                    hasUpdatedPushNoti = NO;
                }
                else {
                    hasUpdatedPushNoti = YES;
                }
                
                if (hasUpdatedUserInfo == NO || hasUpdatedPushNoti == NO) {
                    UIAlertController *vc = [UIAlertController alertControllerWithTitle:[NSBundle sbLocalizedStringForKey:@"ErrorTitle"] message:@"Couldn't update the user information" preferredStyle:UIAlertControllerStyleAlert];
                    UIAlertAction *closeAction = [UIAlertAction actionWithTitle:[NSBundle sbLocalizedStringForKey:@"CloseButton"] style:UIAlertActionStyleCancel handler:nil];
                    [vc addAction:closeAction];
                    
                    self.updatingIndicator.hidden = YES;
                    [self.updatingIndicator stopAnimating];
                    
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self presentViewController:vc animated:YES completion:nil];
                    });
                }
                else {
                    [self dismissViewControllerAnimated:YES completion:nil];
                }
            }];
#else
            if (hasUpdatedUserInfo == NO) {
                UIAlertController *vc = [UIAlertController alertControllerWithTitle:[NSBundle sbLocalizedStringForKey:@"ErrorTitle"] message:@"Couldn't update the user information" preferredStyle:UIAlertControllerStyleAlert];
                UIAlertAction *closeAction = [UIAlertAction actionWithTitle:[NSBundle sbLocalizedStringForKey:@"CloseButton"] style:UIAlertActionStyleCancel handler:nil];
                [vc addAction:closeAction];
                
                self.updatingIndicator.hidden = YES;
                [self.updatingIndicator stopAnimating];
                
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self presentViewController:vc animated:YES completion:nil];
                });
            }
            else {
                [self dismissViewControllerAnimated:YES completion:nil];
            }
#endif
        }];
    }
    else {
        [[UIApplication sharedApplication] unregisterForRemoteNotifications];
        [SBDMain updateCurrentUserInfoWithNickname:[self.nicknameTextField.text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]] profileImage:self.profileImageData completionHandler:^(SBDError * _Nullable error) {
            BOOL hasUpdatedUserInfo;
            if (error != nil) {
                hasUpdatedUserInfo = NO;
            }
            else {
                AFImageDownloader *imageDownloader = [AFImageDownloader defaultInstance];
                NSURLCache *urlCache = imageDownloader.sessionManager.session.configuration.URLCache;
                [urlCache removeAllCachedResponses];
                [imageDownloader.imageCache removeImageWithIdentifier:[SBDMain getCurrentUser].profileUrl];
                
                hasUpdatedUserInfo = YES;
            }
            
#if !(TARGET_OS_SIMULATOR)
            [SBDMain unregisterPushToken:[SBDMain getPendingPushToken] completionHandler:^(NSDictionary * _Nullable response, SBDError * _Nullable error) {
                BOOL hasUpdatedPushNoti;
                
                if (error != nil) {
                    hasUpdatedPushNoti = NO;
                }
                else {
                    hasUpdatedPushNoti = YES;
                }

                if (hasUpdatedUserInfo == NO || hasUpdatedPushNoti == NO) {
                    UIAlertController *vc = [UIAlertController alertControllerWithTitle:[NSBundle sbLocalizedStringForKey:@"ErrorTitle"] message:@"Couldn't update the user information" preferredStyle:UIAlertControllerStyleAlert];
                    UIAlertAction *closeAction = [UIAlertAction actionWithTitle:[NSBundle sbLocalizedStringForKey:@"CloseButton"] style:UIAlertActionStyleCancel handler:nil];
                    [vc addAction:closeAction];
                    
                    self.updatingIndicator.hidden = YES;
                    [self.updatingIndicator stopAnimating];
                    
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self presentViewController:vc animated:YES completion:nil];
                    });
                }
                else {
                    [self dismissViewControllerAnimated:YES completion:nil];
                }
            }];
#else
            if (hasUpdatedUserInfo == NO) {
                UIAlertController *vc = [UIAlertController alertControllerWithTitle:[NSBundle sbLocalizedStringForKey:@"ErrorTitle"] message:@"Couldn't update the user information" preferredStyle:UIAlertControllerStyleAlert];
                UIAlertAction *closeAction = [UIAlertAction actionWithTitle:[NSBundle sbLocalizedStringForKey:@"CloseButton"] style:UIAlertActionStyleCancel handler:nil];
                [vc addAction:closeAction];
                
                self.updatingIndicator.hidden = YES;
                [self.updatingIndicator stopAnimating];
                
                dispatch_async(dispatch_get_main_queue(), ^{
                    [self presentViewController:vc animated:YES completion:nil];
                });
            }
            else {
                [self dismissViewControllerAnimated:YES completion:nil];
            }
#endif
        }];
    }
}

- (void)clickProfileImage {
    UIImagePickerController *mediaUI = [[UIImagePickerController alloc] init];
    mediaUI.sourceType = UIImagePickerControllerSourceTypePhotoLibrary;
    NSMutableArray *mediaTypes = [[NSMutableArray alloc] initWithObjects:(NSString *)kUTTypeImage, nil];
    mediaUI.mediaTypes = mediaTypes;
    [mediaUI setDelegate:self];
    [self presentViewController:mediaUI animated:YES completion:nil];
}

#pragma mark - UIImagePickerControllerDelegate
- (void)imagePickerController:(UIImagePickerController *)picker didFinishPickingMediaWithInfo:(NSDictionary<NSString *, id> *)info
{
    NSString *mediaType = [info objectForKey: UIImagePickerControllerMediaType];
    __weak UserProfileViewController *weakSelf = self;
    [picker dismissViewControllerAnimated:YES completion:^{
        UserProfileViewController *strongSelf = weakSelf;
        if (CFStringCompare ((CFStringRef) mediaType, kUTTypeImage, 0) == kCFCompareEqualTo) {
            NSURL *imagePath = [info objectForKey:@"UIImagePickerControllerReferenceURL"];
            PHAsset *asset = [[PHAsset fetchAssetsWithALAssetURLs:@[imagePath] options:nil] lastObject];
            PHImageRequestOptions *options = [[PHImageRequestOptions alloc] init];
            options.synchronous = YES;
            options.networkAccessAllowed = NO;
            options.deliveryMode = PHImageRequestOptionsDeliveryModeHighQualityFormat;
            [[PHImageManager defaultManager] requestImageDataForAsset:asset options:options resultHandler:^(NSData * _Nullable imageData, NSString * _Nullable dataUTI, UIImageOrientation orientation, NSDictionary * _Nullable info) {
                NSNumber *isError = [info objectForKey:PHImageErrorKey];
                NSNumber *isCloud = [info objectForKey:PHImageResultIsInCloudKey];
                if ([isError boolValue] || [isCloud boolValue] || !imageData) {
                    // fail
                } else {
                    // success, data is in imageData
                    [strongSelf cropImage:imageData];
                }
            }];
        }
    }];
}

- (void)imagePickerControllerDidCancel:(UIImagePickerController *)picker
{
    [picker dismissViewControllerAnimated:YES completion:^{
    }];
}

- (void)cropImage:(NSData *)imageData {
    UIImage *image = [UIImage imageWithData:imageData];
    RSKImageCropViewController *imageCropVC = [[RSKImageCropViewController alloc] initWithImage:image];
    imageCropVC.delegate = self;
    imageCropVC.cropMode = RSKImageCropModeSquare;
    [self presentViewController:imageCropVC animated:NO completion:nil];
}

#pragma mark - RSKImageCropViewControllerDelegate
// Crop image has been canceled.
- (void)imageCropViewControllerDidCancelCrop:(RSKImageCropViewController *)controller
{
    [controller dismissViewControllerAnimated:NO completion:nil];
}

// The original image has been cropped.
- (void)imageCropViewController:(RSKImageCropViewController *)controller
                   didCropImage:(UIImage *)croppedImage
                  usingCropRect:(CGRect)cropRect
{
    self.profileImageView.image = croppedImage;
    [controller dismissViewControllerAnimated:NO completion:nil];
}

// The original image has been cropped. Additionally provides a rotation angle used to produce image.
- (void)imageCropViewController:(RSKImageCropViewController *)controller
                   didCropImage:(UIImage *)croppedImage
                  usingCropRect:(CGRect)cropRect
                  rotationAngle:(CGFloat)rotationAngle
{
    self.profileImageView.image = croppedImage;
    self.profileImageData = UIImageJPEGRepresentation(croppedImage, 1);
    [controller dismissViewControllerAnimated:NO completion:nil];
}

// The original image will be cropped.
- (void)imageCropViewController:(RSKImageCropViewController *)controller
                  willCropImage:(UIImage *)originalImage
{
    // Use when `applyMaskToCroppedImage` set to YES.
}

@end

File: syncmanager/swift/Sample-SyncManager-iOS-swift/UserProfile/UserProfileViewController.swift
//
//  UserProfileViewController.swift
//  SendBird-iOS-LocalCache-Sample-swift
//
//  Created by Jed Gyeong on 6/21/17.
//  Copyright © 2017 SendBird. All rights reserved.
//

import UIKit
import SendBirdSDK
import RSKImageCropper
import Alamofire
import AlamofireImage
import Photos
import MobileCoreServices

class UserProfileViewController: UIViewController, UIImagePickerControllerDelegate, UINavigationControllerDelegate, RSKImageCropViewControllerDelegate {

    @IBOutlet weak var navItem: UINavigationItem!
    @IBOutlet weak var profileImageView: UIImageView!
    @IBOutlet weak var nicknameTextField: UITextField!
    @IBOutlet weak var bottomMargin: NSLayoutConstraint!
    @IBOutlet weak var pushNotificationSwitch: UISwitch!
    @IBOutlet weak var updatingIndicator: UIActivityIndicatorView!
    
    var profileImageData: Data?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
        let negativeRightSpacer = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.fixedSpace, target: nil, action: nil)
        negativeRightSpacer.width = -2
        let rightDisconnectItem = UIBarButtonItem(title: "Save", style: UIBarButtonItem.Style.plain, target: self, action: #selector(save))
        rightDisconnectItem.setTitleTextAttributes([NSAttributedString.Key.font: Constants.navigationBarButtonItemFont()], for: UIControl.State.normal)
        
        let negativeLeftSpacer = UIBarButtonItem(barButtonSystemItem: UIBarButtonItem.SystemItem.fixedSpace, target: nil, action: nil)
        negativeLeftSpacer.width = -2
        let leftProfileItem = UIBarButtonItem(title: "Close", style: UIBarButtonItem.Style.plain, target: self, action: #selector(save))
        leftProfileItem.setTitleTextAttributes([NSAttributedString.Key.font: Constants.navigationBarButtonItemFont()], for: UIControl.State.normal)

        self.navItem.rightBarButtonItems = [negativeRightSpacer, rightDisconnectItem]
        self.navItem.leftBarButtonItems = [negativeLeftSpacer, leftProfileItem]
        
        self.profileImageView.af_setImage(withURL: URL(string: (SBDMain.getCurrentUser()?.profileUrl!)!)!)
        
        self.profileImageData = nil
        
        self.updatingIndicator.isHidden = true
        
        let profileImageTapRecognizer = UITapGestureRecognizer(target: self, action: #selector(clickProfileImage))
        self.profileImageView.isUserInteractionEnabled = true
        self.profileImageView.addGestureRecognizer(profileImageTapRecognizer)
        
        self.nicknameTextField.text = SBDMain.getCurrentUser()?.nickname
        
        let isRegisteredForRemoteNotifications = UIApplication.shared.isRegisteredForRemoteNotifications
        self.pushNotificationSwitch.isOn = isRegisteredForRemoteNotifications
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func close() {
        self.dismiss(animated: false, completion: nil)
    }
    
    @objc func save() {
        if self.nicknameTextField.text?.trimmingCharacters(in: CharacterSet.whitespaces).count == 0 {
            return
        }
        
        self.updatingIndicator.isHidden = false
        self.updatingIndicator.startAnimating()
        
        let isRegisteredForRemoteNotifications = self.pushNotificationSwitch.isOn
        if isRegisteredForRemoteNotifications {
            UIApplication.shared.registerForRemoteNotifications()
            SBDMain.updateCurrentUserInfo(withNickname: self.nicknameTextField.text?.trimmingCharacters(in: CharacterSet.whitespaces), profileImage: self.profileImageData, completionHandler: { (error) in
                var hasUpdatedUserInfo: Bool?
                if error != nil {
                    hasUpdatedUserInfo = false
                }
                else {
                    let imageDownloader = UIImageView.af_sharedImageDownloader
                    let urlCache = imageDownloader.sessionManager.session.configuration.urlCache
                    urlCache?.removeAllCachedResponses()
                    _ = imageDownloader.imageCache?.removeImage(withIdentifier: (SBDMain.getCurrentUser()?.profileUrl)!)
                    
                    hasUpdatedUserInfo = true
                }
                
#if !(arch(i386) || arch(x86_64))
                SBDMain.registerDevicePushToken(SBDMain.getPendingPushToken()!, unique: true, completionHandler: { (status, error) in
                    var hasUpdatedPushNoti: Bool?
                    
                    if error != nil {
                        hasUpdatedPushNoti = false
                    }
                    else {
                        hasUpdatedPushNoti = true
                    }
                    
                    if hasUpdatedUserInfo == false || hasUpdatedPushNoti == false {
                        let vc = UIAlertController(title: "Error", message: error?.domain, preferredStyle: UIAlertControllerStyle.alert)
                        let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                        vc.addAction(closeAction)
                        DispatchQueue.main.async {
                            self.present(vc, animated: true, completion: nil)
                        }
                        
                        self.updatingIndicator.isHidden = true
                        self.updatingIndicator.stopAnimating()
                        
                        DispatchQueue.main.async {
                            self.present(vc, animated: true, completion: nil)
                        }
                    }
                    else {
                        self.dismiss(animated: true, completion: nil)
                    }
                })
#else
                if hasUpdatedUserInfo == false {
                    let vc = UIAlertController(title: "Error", message: error?.domain, preferredStyle: UIAlertController.Style.alert)
                    let closeAction = UIAlertAction(title: "Close", style: UIAlertAction.Style.cancel, handler: nil)
                    vc.addAction(closeAction)
                    DispatchQueue.main.async {
                        self.present(vc, animated: true, completion: nil)
                    }
                    
                    self.updatingIndicator.isHidden = true
                    self.updatingIndicator.stopAnimating()
                    
                    DispatchQueue.main.async {
                        self.present(vc, animated: true, completion: nil)
                    }
                }
                else {
                    self.dismiss(animated: true, completion: nil)
                }
#endif
            })
        }
        else {
            UIApplication.shared.unregisterForRemoteNotifications()
            SBDMain.updateCurrentUserInfo(withNickname: self.nicknameTextField.text?.trimmingCharacters(in: CharacterSet.whitespaces), profileImage: self.profileImageData, completionHandler: { (error) in
                var hasUpdatedUserInfo: Bool?
                if error != nil {
                    hasUpdatedUserInfo = false
                }
                else {
                    let imageDownloader = UIImageView.af_sharedImageDownloader
                    let urlCache = imageDownloader.sessionManager.session.configuration.urlCache
                    urlCache?.removeAllCachedResponses()
                    _ = imageDownloader.imageCache?.removeImage(withIdentifier: (SBDMain.getCurrentUser()?.profileUrl)!)
                    
                    hasUpdatedUserInfo = true
                }
                
#if !(arch(i386) || arch(x86_64))
                SBDMain.unregisterPushToken(SBDMain.getPendingPushToken()!, completionHandler: { (response, error) in
                    var hasUpdatedPushNoti: Bool?
                    
                    if error != nil {
                        hasUpdatedPushNoti = false
                    }
                    else {
                        hasUpdatedPushNoti = true
                    }
                    
                    if hasUpdatedUserInfo == false || hasUpdatedPushNoti == false {
                        let vc = UIAlertController(title: "Error", message: error?.domain, preferredStyle: UIAlertControllerStyle.alert)
                        let closeAction = UIAlertAction(title: "Close", style: UIAlertActionStyle.cancel, handler: nil)
                        vc.addAction(closeAction)
                        DispatchQueue.main.async {
                            self.present(vc, animated: true, completion: nil)
                        }
                        
                        self.updatingIndicator.isHidden = true
                        self.updatingIndicator.stopAnimating()
                        
                        DispatchQueue.main.async {
                            self.present(vc, animated: true, completion: nil)
                        }
                    }
                    else {
                        self.dismiss(animated: true, completion: nil)
                    }
                })
#else
                if hasUpdatedUserInfo == false {
                    let vc = UIAlertController(title: "Error", message: error?.domain, preferredStyle: UIAlertController.Style.alert)
                    let closeAction = UIAlertAction(title: "Close", style: UIAlertAction.Style.cancel, handler: nil)
                    vc.addAction(closeAction)
                    DispatchQueue.main.async {
                        self.present(vc, animated: true, completion: nil)
                    }
                    
                    self.updatingIndicator.isHidden = true
                    self.updatingIndicator.stopAnimating()
                    
                    DispatchQueue.main.async {
                        self.present(vc, animated: true, completion: nil)
                    }
                }
                else {
                    self.dismiss(animated: true, completion: nil)
                }
#endif
            })
        }
    }
    
    @objc func clickProfileImage() {
        let mediaUI = UIImagePickerController()
        mediaUI.sourceType = UIImagePickerController.SourceType.photoLibrary
        let mediaTypes = [String(kUTTypeImage)]
        mediaUI.mediaTypes = mediaTypes
        mediaUI.delegate = self
        self.present(mediaUI, animated: true, completion: nil)
    }
    
    //MARK: UIImagePickerControllerDelegate
    func imagePickerController(_ picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [UIImagePickerController.InfoKey : Any]) {
        picker.dismiss(animated: true) {
            let theMediaType: String? = info[UIImagePickerController.InfoKey.mediaType] as? String
            guard let mediaType: String = theMediaType else {
                return
            }
            
            if Utils.isKindOfImage(mediaType: mediaType) {
                let theAsset: PHAsset? = info[UIImagePickerController.InfoKey.phAsset] as? PHAsset
                guard let asset: PHAsset = theAsset else {
                    return
                }
                
                let options: PHImageRequestOptions = PHImageRequestOptions()
                options.isSynchronous = true
                options.isNetworkAccessAllowed = false
                options.deliveryMode = PHImageRequestOptionsDeliveryMode.highQualityFormat
                PHImageManager.default().requestImageData(for: asset, options: options, resultHandler: { (theImageData, dataUTI, orientation, info) in
                    let isError: Bool? = info?[PHImageErrorKey] as? Bool
                    let isCloud: Bool? = info?[PHImageResultIsInCloudKey] as? Bool
                    
                    guard let hasError: Bool = isError, hasError == false else {
                        return
                    }
                    guard let hasCloud: Bool = isCloud, hasCloud == false else {
                        return
                    }
                    guard let imageData: Data = theImageData else {
                        return
                    }
                    
                    self.cropImage(imageData: imageData)
                })
            }
        }
    }
    
    func imagePickerControllerDidCancel(_ picker: UIImagePickerController) {
        picker.dismiss(animated: true, completion: nil)
    }
    
    func cropImage(imageData: Data) {
        let image = UIImage(data: imageData)
        let imageCropVC: RSKImageCropViewController = RSKImageCropViewController(image: image!)
        imageCropVC.delegate = self
        imageCropVC.cropMode = RSKImageCropMode.square
        self.present(imageCropVC, animated: false, completion: nil)
    }
    
    //MARK: RSKImageCropViewControllerDelegate
    // Crop image has been canceled.
    func imageCropViewControllerDidCancelCrop(_ controller: RSKImageCropViewController) {
        controller.dismiss(animated: false, completion: nil)
    }
    
    // The original image has been cropped.
    func imageCropViewController(_ controller: RSKImageCropViewController, didCropImage croppedImage: UIImage, usingCropRect cropRect: CGRect) {
        self.profileImageView.image = croppedImage
    }
    
    // The original image has been cropped. Additionally provides a rotation angle used to produce image.
    func imageCropViewController(_ controller: RSKImageCropViewController, didCropImage croppedImage: UIImage, usingCropRect cropRect: CGRect, rotationAngle: CGFloat) {
        self.profileImageView.image = croppedImage
        self.profileImageData = croppedImage.jpegData(compressionQuality: 1)
        controller.dismiss(animated: false, completion: nil)
    }
    
    // The original image will be cropped.
    func imageCropViewController(_ controller: RSKImageCropViewController, willCropImage originalImage: UIImage) {
        // Use when `applyMaskToCroppedImage` set to YES.
    }
}
